Add blog post on the new rr features (#764)
JuliaLang/www.julialang.org: blog/2020/05/rr.md

                    theme="asciinema"
                    title="Recording"
                    src="/assets/blog/2020-05-02-rr/record.cast">
  </asciinema-player>
  <asciinema-player author="Keno Fischer"
                    cols="100"
                    id="replay_player"
                    class="hidden"
                    rows="40"
                    speed="1"
                    theme="asciinema"
                    title="Recording"
                    src="/assets/blog/2020-05-02-rr/replay.cast">
  </asciinema-player>
  </div>
  <image id="arch_diagram" src="/assets/blog/2020-05-02-rr/arch.svg"/>
  </div>
  <script>
    document.getElementById('record_player').addEventListener('ended', function(e) {
      document.getElementById('player_main').classList.add("animated");
      document.getElementById('arch_diagram').classList.add("animated");
      document.getElementById('player_main').addEventListener('animationend', () => {
        document.getElementById('record_player').classList.add("hidden");
        document.getElementById('replay_player').classList.remove("hidden");
        document.getElementById('replay_player').play();
      });
    });
  </script>
~~~

> 

Author: Good suggestion, thanks.

feel the effort to write a high-quality bug report is too high, since their
specific setup would be too complicated to describe and reproduce. Such
experiences are frustrating both for the users encountering them and for us,
because we don't learn about them and sometimes only hear about it years later,
when we get told that somebody gave up on their project in Julia because they
encountered some crashing problem that they didn't feel qualified to reduce into
a concise bug report and subsequently gave up on using Julia for their project.
Lastly, we don't want or expect our users to be expert bug reporters. They are
often working scientists who are good programmers, but may not have any
background in software engineering. These are some of our most valuable users,
and we want to make sure their bugs are addressed.

For users who did encounter such particularly difficult problems, we have for
a long time had one particular answer: If you can reproduce it on a linux
machine and get us a trace from the `rr` tool <https://rr-project.org/>, we can
probably get it fixed for you very rapidly. For the uninitiated, `rr` is a Linux

> debugging tool originally written by Robert O'Callahan at Mozilla. Is it a tool

Reviewer: ```suggestion
debugging tool originally written by Robert O'Callahan at Mozilla. It is a tool
```

> debugging tool originally written by Robert O'Callahan at Mozilla. It is a tool

Reviewer 2: This is a bit misleading as written, I didn't write much actual rr code before it became a usable tool :-). Maybe write "originally developed at Mozilla by Robert O'Callahan and others"?

Author: Thanks, I wasn't quite aware of the history, applied as suggested.

known as a "time traveling debugger" or "reverse execution engine". Essentially,
`rr` splits reproducing bugs into a "record" and a "replay" phase. The record
phase is performed by the bug reporter. During this phase, `rr` creates a perfect
record of the execution, including the *bitwise exact* memory and register state
after every instruction. This trace can then be analyzed during the replay
phase (which can be performed by a different developer on a different machine).
Capabilities like these have long been imagined in academia, but it is extremely
hard to achieve in a way that does not introduce large overheads or distort

> regular execution. `rr` is that first such tool that (in our experience) is

Reviewer 3: ```suggestion
regular execution. `rr` is the first such tool that (in our experience) is
```

performant enough to be used in the regular course of development. It is worth
discussing how this is achieved (and what limitations the approach has), but
first let's take the capabilities for granted and take a look at the workflow
it enables.

In Julia 1.5, there is now a new command line flag `--bug-report=rr` that will
automatically create and upload an `rr` recording. An example usage is shown
in the animation at the start of this post (which just creates a deliberate
crash by unsafely dereferencing a bad pointer). However to summarize:

> 1. The bug reporter passes `--bug-report=rr` to her julia instance and

Reviewer: ```suggestion
1. The bug reporter passes `--bug-report=rr` to their julia instance and
```

Author: This was a deliberate editorial choice

   reproduces whatever bug they are attempting to reproduce.

2. Once julia exits or crashes, the bug reporter is prompted to authorize an
   upload by clicking on a link (GitHub-based authentication is used for abuse

>    protection). She then gets back a link to include when filing a bug report

Author: As above

   against julia or some other package.

3. Any developer can use the link to obtain the recording and analyze it on
   their own machine.

In addition to this manual mechanism, we are also switching our linux CI systems
to automatically create `rr` traces of any execution. That way, if a CI run fails,
we are guaranteed to be able to debug it.

If a bug report includes a link to an `rr` trace, no further reproduction
instructions are required (if the bug is something non-obvious like
unexpected behavior, some comments on what the expected behavior was may

> still be helpful), since the `rr` trace is guaranteed to perfectly capture

Reviewer 4: I think I'd still want the bug report to still contain the usual suspects (what was the expected result and unexpected observation). Could be helpful in case the `rr` trace is unnecessary (clear from the final backtrace, for example) or breaks or just needs triage, and so on.

Author: Well, part of the point is that it should be possible to automatically extract all the useful information from the trace and post it automatically - versioninfo, manifests, etc, even if you don't want to do a full replay. https://github.com/JuliaLang/julia/issues/35601 was a MVP of that.

the environment the bug was reproduced in. This almost immediately knocks
out all the common problems I started this post with. "Works for me" is no
longer an available answer. If it's in the trace, it broke on somebody's machine
and can be debugged. "Heisenbug"s are no longer an issue. If it was captured
once in `rr`, it can always be debugged. It even solves the busy-expert problem,
as it allows non-experts to help out with triage. If such a report does not
contain an `rr` trace, any developer, and in particular non-experts can attempt to
reproduce the bug and create one. Even if the expert is still required to do
the final diagnosis, doing so from an `rr` trace is orders of magnitude faster
than from a plain bug report.

# Chronomancy for dummies

A computer is fundamentally a deterministic machine. Given equivalent
states as inputs, most instructions will produce a deterministic state

> as output. So where do all the subtle execution differences comes from that

Reviewer 3: ```suggestion
as output. So where do all the subtle execution differences come from that
```

produce diverging executions and prevent bugs from reproducing? Well, the

> complete answer is complicated an has lots of details, but is roughly split

Reviewer: ```suggestion
complete answer is complicated and has lots of details, but is roughly split
```

up into:

1) The input state is large. Probably the state of your entire hard drive and
    memory. That's at least `2^2^large` bits of state right there each of which
    could potentially cause a difference in execution. You really want to figure
    out what state is relevant, since `2^2^large` is probably too many bits to send
    somebody.

2) Any input or other asynchronous events. This can mean user input, or
    data from the network (via the NICs) or other devices. It also means things like
    the timing of interrupts.

3) Execution ordering and data races in multi-threaded executions.

> 

Reviewer 4: ```suggestion
4) Direct observations of non-deterministic noise: For example, cache timing attacks (ala Spectre), reading `/dev/random`, or instructions such as `RDRAND`.
```

Author: I didn't introduce the OS here yet, so I'll drop /dev/random, but include the other two.

However, in theory if a tool was able to capture 100% of the relevant state
from these categories, it could repeated generate exactly the same memory image.
This is not a novel idea, but the devil is in the detail. As already mentioned,
just capturing everything is probably too large, if at all possible (e.g. how

> do you capture interrupt timing - how do you even talk about time in this setting?).

Reviewer 4: ```suggestion
do you capture interrupt timing - how do you even talk about time in this setting?).
```

We could lie about the rate of passage of time to it?

The other possible problem here is that multithreading lies outside of the total input state space (cannot be enumerated by a Turing machine).

Author: The point I wanted to make was slightly different. It's that the notion of time in an execution is hard to define in such a way that makes it useful. Real time is useless, because instruction issue frequency is not anywhere near to constant. Executed instructions are fine, but may not be available in hardware etc. I should probably just say that explicitly.

> In other words, for some input state $\mathcal{S_{i}}$,

Reviewer 4: "We can model this with an equation."?? Not really words 😂

the output state $\mathcal{S_{i+1}}$ is determined by some pure, deterministic
function $\mathcal{I}_{ip(\mathcal{S}_i)}$, corresponding to the instruction to be
executed at time $i$, or some asynchronous event $\mathcal{A}_i$

$$ \mathcal{S}_{i+1} = \begin{cases}

>     \mathcal{A}_i(S_{i}) & \text{if an asynchronous even occured at time $i$} \\

Reviewer 3: ```suggestion
    \mathcal{A}_i(S_{i}) & \text{if an asynchronous event occured at time $i$} \\
```

    \mathcal{I}_{\text{ip}(\mathcal{S}_i)} & \text{otherwise}
\end{cases} $$

> Did that help? No? oh. Well, it made me feel better about my physics degree and

Reviewer 4: Possible alternative wording, perhaps:
```suggestion
Assuming we can discretize the asynchronous event into..... Okay, I realize this probably isn't helping. But I've been told it's not research until it has at least one equation in it and it made me feel better about my physics degree. Also, people worked really hard to get that nice equation rendering working, so I better use it. Anyway, where were we?
```

Author: Doesn't quite sound like me ;). The purpose of this joke is to break up the dense technical content with some levity and personal connection to the reader - as such I'd like to keep it in my voice.

also I've been told it's not research until it has at least one equation in it,
so there you go. Also, people worked really hard to get the equation rendering working,

> on this blog, so I better use it. Anyway, were were we?

Reviewer 5: ```suggestion
on this blog, so I better use it. Anyway, where were we?
```

> Ah yes, `rr` works by treating the kernel/userspace boundary as the determinism boundary.

Reviewer 4: Not sure that immediately helps clarify anything either, for someone who's never thought about a "kernel".
```suggestion
Ah yes, to make this equation deterministic, `rr` only needs to track any data that came from an asynchronous event or outside data source. And this actually isn't much data. Conveniently, the kernel/userspace boundary already is splitting our program into the majority of the deterministic computational work (our program!) and all unknown input state (files, networking, time). If a part...
```

Author: Fair. I'll attempt to clarify this paragraph for that audience.

This has a lot of advantages. For one, the kernel isolates a lot of state. If a
part of the disk state (e.g. a file) wasn't requested through the kernel, you
can be guaranteed that it didn't affect the process state (assuming the kernel
works correctly of course). It abstracts over hardware details and provides
a uniform interface to resources like the network. It also isolates processes
from another, so if there is one process of interest (e.g. julia) only the
activity relevant to it has to be recovered.

> Of course, there are some drawbacks as well. In order to function `rr` has to

Reviewer 4: ```suggestion
Even with that knowledge, this is not easy to accomplish. In order to function, `rr` must
```

Author: Applied with a slight reword

have an extremely precise model of the operation of the kernel and the ways

> in which it interacts with userspace. On Linux this interface is supposedly

Reviewer 4: ```suggestion
in which it interacts with userspace. The Linux developers work hard to attempt to keep this interface
```

Author: Applied.

stable, but few applications put as stringent a constraint on that promise as

> `rr` does (e.g. even a single bit difference can be problematic).

Reviewer 4: ```suggestion
`rr` does. Sometimes, even a single bit difference can be problematic!
```

Author: Applied!

> The details here are quite involved, and I recommend reading Robert's

Reviewer 4: ```suggestion
If these details interest you, I recommend reading Robert's
```

Author: Applied.

[technical report](https://arxiv.org/abs/1610.02144) for a deeper overview of
some of `rr`'s design points (though even that only scratches the surface).

> One final though for this section is the origin of the term "time-traveling"

Reviewer: ```suggestion
One final thought for this section is the origin of the term "time-traveling"
```

> One final thought for this section is the origin of the term "time-traveling"

Reviewer 4: Apparently the only use of this term?
```suggestion
This feature of the `rr` tool is called "time-traveling" debugging.
```

Author: Applied.

Author: I'm pretty sure the term predates rr by quite a bit.

debugger. It stems from the mode of analysis that a system like this enables.
In a traditional debugger, it is possible to execute one instruction at a time,
and step from memory state to memory state in the forwards direction. Systems

> like `rr` allow the opposite during replay: Step backwards through the state of

Reviewer 4: ```suggestion
like `rr` allow the opposite during replay: step backwards through the state of
```

Author: applied

the system. Under the hood this is accomplished by playing forward from the
beginning (or more likely some intermediate checkpoint created for performance
reasons), until the previous state is reached. However, to the end user, the
illusion of going backwards in time is presented and an extremely useful mental
model for debugging.

# Performance considerations

While the previous section described the operating principle of `rr`, it does
not do justice to the reason why `rr` works so well. That reason is simple:
performance. Overhead for recording of single threaded processes
is generally below 2x, most often between 2% and 50% (lower for purely
numerical calculations, higher for workloads that interact with the OS).

> Recording multi-threaded processes is tougher. By default `rr` serializes

Reviewer 4: ```suggestion
Recording multiple threads or processes is harder. Instead, `rr` serializes the execution of each thread. This incurs overhead linear in the number of threads. But, in generality, this is required for reproducibility. It is thus a good idea to try and reproduce any issues at low core count. There is interesting academic research in other techniques, but nothing is close to production-ready yet.
```
(and kill the rest of this paragraph, or at least save it for your JuliaCon presentation)

Author: Fair. Applied in spirit, but with slight rewordings for flow.

execution of tasks that access the same shared memory. In generality this is
required for correctness, since it is not possible to observe and record the
interleaving of memory operations to shared memory spaces. However, there is
hope. If the multi-threaded program is well-behaved (or mostly well-behaved)
in that communication happens by message passing rather than shared memory, or
if accesses to shared memory are explicitly synchronized in a recordable fashion,
multi-threaded recording becomes possible. `rr` currently supports the former,
if the messages are passed via operating system pipes. We are interested in
techniques for the latter and there is some interesting academic work in this
direction, but nothing close to production ready. A complete solution would
probably combine compiler-based analysis of safe regions with dynamic
enforcement of safety for regions where this cannot be proven by the compiler.
However, that is for future work. For the moment, recording multi-threaded,
shared-memory applications will likely incur overhead linear in the number of
threads. It is thus a good idea to try and reproduce any issues at low core count.

> With that caveat in mind. Here is some real benchmark numbers of overhead for

Reviewer 4: ```suggestion
With that caveat in mind, here is some real benchmark numbers of overhead for
```

Author: Applied

> the julia test suite. The tests suite mostly uses message-passing based

Reviewer 4: ```suggestion
the julia test suite. The test suite mostly uses message-passing based
```

Author: applied

parallelism for running multiple tests in parallel. Only the `threads` test

> incurs the shared memory overhead penalty. Here are the measurements:

Reviewer 4: ```suggestion
incurs the shared memory overhead penalty, so most tests run with little penalty. Here are the measurements to see for yourself:
```

Author: Outdated after re-work to add the benchmark results

> [data]

Reviewer 4: [placeholder comment]

> The total compressed size of a trace for a full run of the test suite (about

Reviewer 4: ```suggestion
Surprisingly too, your traces may often be quite small! The total compressed size of a trace for a full run of the test suite (about
```

Author: Fair, people may not have accurate calibrated expectations ;). Applied.

30 minutes on 10 cores) is about three gigabytes.

# Hardware and Software limitations

As was already mentioned, `rr` currently only works on Linux. However, there
are more restrictions. At the moment only x86 chips with Intel microarchitecture
are supported. `rr` relies on hardware performance counters to establish a
precise and accurate notion of time for asynchronous events. If these hardware
counters are not available or not precise enough, recording using `rr` will not
work. We and others have investigated whether it would be possible to port rr
to other architectures. To the best of my knowledge, the current thinking on
this is as follows:

- For x86_64 with AMD Ryzen microarchitecture, it seems potentially possible, but the
    AMD performance counters are less accurate than those found on Intel chips.
    It may be possible to compensate for this in software, but the in-accuracies
    are ill-understood. Help in investigation is welcome ([issue](https://github.com/mozilla/rr/issues/2034)).

> - For Aarch64, ll/sc-based atomics introduce additional problems, since various

Reviewer 4: ```suggestion
- For AArch64, ll/sc-based atomics introduce additional problems, since various
```

Author: applied here and elsewhere.

    external factors (interrupts, descheduling, etc.) can be observed as sc aborts.
    With proper hardware support, this would be possible to work around, but such

>     hardware support appeared to be too in-accurate on older generation Aarch64 chips

Reviewer 4: ```suggestion
    hardware support appeared to be inaccurate on older generation AArch64 chips
```

Author: applied

>     and have been removed on recent generation ones. That said, Aarch64

Reviewer 4: ```suggestion
    and has been removed on recent generation ones. That said, AArch64
```

Author: applied

    microarchitectures are varied and as evidenced by the x86_64 experience, these

>     things can vary quite a bit by microarchitecture. Investigation of the hardware

Reviewer 4: ```suggestion
    the quality of implementation can vary quite a bit by microarchitecture. Investigation of the hardware
```

Author: applied with tweaks

    capabilities of various Aarch64 microarchitectures would be very welcome
    ([issue](https://github.com/mozilla/rr/issues/1373)).

> - We have looked into rr on POWER9, which is found in several large supercomputers

Reviewer 4: ```suggestion
- We have looked into `rr` on POWER9, which is found in several large supercomputers
```

Author: applied

    that run julia jobs. POWER9 has similar issues to Aarch64 since it is also an
    ll/sc architecture, as well as supporting hardware transactional memory. In theory,
    we believe the hardware exists to compensate for these issues, but preliminary
    investigation suggests the hardware is not accurate enough to support `rr`. That

>     said, I understand there's active discussions with IBM to determine whether this

Reviewer 4: How strongly do you want to stand behind this rumor?

Author: Strongly enough such that if anybody from the IBM hardware side reads this, they'll know to ask around internally and find the right people to get in touch with us.

    analysis is correct.

> One additional point of complication, is that it is quite easy for hypervisors

Reviewer 4: ```suggestion
One additional point of complication, in our experience, it is quite common for hypervisors to disable the CPU capabilities required by `rr`, rendering it incapable of functioning inside virtual machines.
```

Author: applied

(such as those used by cloud vendors) to interfere with the CPU capabilities
required by `rr`. My understanding is that `rr` works fine on the latest generation

> of AWS machines, but not on GCP or Azure.

Reviewer 4: Possibly consider spelling out these acronyms (or at least give company names)

Author: Fair

> of AWS machines, but not on GCP or Azure.

Reviewer 2: You can confidently state that rr does not work on GCP or Azure but does work on AWS instances where hardware performance counters are enabled. (E.g. AWS C-instances where the VM occupies a whole socket, e.g. c5.9xlarge and bigger ... though those details don't matter for this post.)

Author: Done.

Lastly, we shouldn't fail to mention GPUs. GPUs are heavily utilized by Julia
users, but are currently not supported by `rr`. Changing this is tricky. GPUs
often allow direct memory access between the device and the userspace program
being recorded. If the interface to the GPU is known, this can definitely be
mitigated, potentially at some performance cost, e.g. by double buffering. For
GPUs with open source drivers, e.g. AMDGPU, there is active community interest
in building such a solution. It is not an easy task, but I think it has a
decent chance of succeeding, since open source drivers allow insight into exactly
how the GPU works and when it will modify process memory. GPUs with proprietary
drivers are a lot harder. For one, their interface to userspace is not necessarily
known. Additionally, proprietary drivers tend to be a lot less well behaved

> than open source ones (the linux kernel review process tends to filter out

Reviewer 4: ```suggestion
than open source ones (the Linux kernel review process tends to filter out
```

Author: Applied.

`rr` much faster), but without a significant user base of such features they
have little incentive to care. I'm hoping that by rolling out these capabilities
widely, these incentives will start presenting themselves. Judging from the size
of the community and discounting for those that use unsupported hardware or
operating system, the number of users for whom this feature will be available
like numbers in the hundreds of thousands. That isn't super large, by hardware
vendor standards, but it isn't trivially small either. If any hardware vendor
is interested in making this work, we'd be happy to talk to you (and as I
mentioned, some of this is already in progress).

# A word on privacy

By its nature, an `rr` trace will contain any file touched by the process during
its lifetime. In particular, this may contain things like your julia history,
configuration files for your system, any secrets entered or read (i.e. make
sure to use ssh-agent if your reproducer involves using SSH for authentication),
any private code you may be using, etc. We are investigating building tooling
to help understand what's in the trace and anonymize parts of the trace that are

> potentially sensitive, but do not have long-ranging effects in the trace. We

Reviewer 4: What's "long-ranging effects"?

Author: Even if you don't read something from your history file, it could become unallocated garbage that later is used as uninitialized memory and causes a problem. I'll spell out this example.

Author: Actually, I think that's too detailed, I'll just say do not affect the trace

> potentially sensitive, but do not have long-ranging effects in the trace. We

Reviewer 2: You may already know about it but there's a little bit of prior art for this in https://github.com/Pernosco/pernosco-submit/blob/be4f5067c90d5c67cb5068d5d8f586875f0e2260/pernosco-submit#L239.

Author: I hadn't seen that. Seems like a useful thing that we should do and doesn't require a lot of effort.

> will also likely be disabling history by default if the `--bug-report` command

Reviewer 4: Let's just do this, then update the blog to mention it isn't included

Author: https://github.com/JuliaLang/BugReporting.jl/commit/b4a9580a71f833c976ec310fbe22ce2df6c8345e

Author: Will be part of the new tag (along with a few other bug fixed and switching the compression format).
